DESIGN.md

```diff
@@ -0,0 +1,167 @@
+get information about their progress (via logs and exit code),
+and allow to stop the processes if needed.
+
+# Worker
+
+`worker` is a Rust library that provides an async API
+to spawn, control and inspect the processes.
+It aims to provide all the control functionality
+required by `paasd`, without tying it to the external API specifics.
+This means it does not provide the gRPC protocol or transport.
+Otherwise, functions provided by `worker` map almost 1-to-1
+to endpoints provided by `paasd`.
+
+# Paasd
+
+`paasd` is a daemon that clients use to manage the processes.
+It utilizes `worker` under the hood and provides a gRPC API
+to access the functionality.
+
+It also stores the information about executing and finished processes in-memory.
+This in-memory store keeps the information for the lifetime of the daemon.
+
+## API
+
+Clients interact with `paasd` via a gRPC API. See `paas.proto` for RPC definition.
+
+### `Exec`
+
+`Exec` takes in the command line as an array of strings.
+The first argument is the executable, the rest are arguments to the executable.
+This is similar to the `argv` argument of `execv*` family.
+
+On successfully spawning the process,
+it is given a UUID, which is returned to the client
+for use in future commands in regards to this process.
+
+### `GetLogs`
+
+`GetLogs` returns a stream of logs produced by the process with the given ID.
+Both stdout and stderr are combined into a single stream.
+`paasd` accumulates all logs produced by the process.
+Upon the request, the stream will return all previous logs
+and will then return messages as they are produced by the process.
+
+`paasd` assumes that logs consist of valid UTF-8 textual data and uses line-based buffering.
+Most Unix environments will use UTF-8
+and most Unix-y utilities will produce textual logs with reasonably sized lines.
+
+`GetLogs` prioritizes latency over throughput.
+It will try to send multiple log lines in a single message, but this is not guaranteed.
+In general, new log messages are sent as soon as possible, without additional buffering.
+
+### `GetStatus`
+
+`GetStatus` returns the status of the process with the given ID.
+Currently, the response only indicates whether the process has exited,
+in which case the response will contain either the signal number
+(if the process has been killed by a signal), or the exit code otherwise.
+
+### `Stop`
+
+`Stop` stops the process with the given ID if it is still running.
+It first tries to let the program terminate gracefully by sending a SIGTERM.
+If the process is still running after a set amount of time, it will send a SIGKILL.
+This mirrors the behavior of many process managers, for example,
+[systemd](https://stackoverflow.com/questions/42978358/how-systemd-stop-command-actually-works).
+
+It is a no-op to issue a `Stop` command for a process that has already been finished.
+However, trying to `Stop` a process that is in the process of being stopped returns an error.
```

Comment:
I think this is a valid approach, but I'm a bit curious why consider a program that is stopping to be an error on the stop API, but stopping an already exited program not an error?

Reply:
While this choice has been made somewhat arbitrarily, I'll try to explain it away.

Let's say both of these scenarios were treated as an error. Then, to avoid the error, the user must "look before they leap":  first get the status of the process, then, if the process is still running, issue a `Stop`. But this is inherently racy, as the process might finish in between the two calls.

As for going the other way and treating both scenarios as a valid call... I guess this would be viable at the service (`paasd`) level. But at the library level, I would really like to enforce that the process must be "stopped" only once. I think if we let the user issue multiple concurrent `Stop`-s, resulting in multiple SIGTERMs, some processes might treat that as a request to shutdown forcefully (which we do not want, unless the timeout lapses).

Also looking more broadly, where `paasd` may be used not only via an interactive command line client, but by some automated script, multiple `Stop`-s in rapid succession would seem to me like a programming error on the client side. However, `Stop` for a process that has already finished does not seem like an inherent bug in the client, because, like mentioned before, "get status then stop" is racy.